src/schedule/dma_multi_chan_domain.c

```diff
@@ -0,0 +1,259 @@
+	int j;
+
+	for (i = 0; i < dma_domain->num_dma; ++i) {
+		for (j = 0; j < dmas[i].plat_data.channels; ++j) {
+			/* channel not set as scheduling source */
+			if (!dma_is_scheduling_source(&dmas[i].chan[j]))
+				continue;
+
+			/* channel not running */
+			if (dmas[i].chan[j].status != COMP_STATE_ACTIVE)
+				continue;
+
+			/* channel owned by different core */
+			if (core != dmas[i].chan[j].core)
+				continue;
+
+			/* channel has been already running */
+			if (dma_domain->channel_mask[i][core] & BIT(j))
+				continue;
+
+			dma_interrupt(&dmas[i].chan[j], DMA_IRQ_CLEAR);
+
+			/* register only if not aggregated or not registered */
+			if (!dma_domain->aggregated_irq ||
+			    !dma_domain->channel_mask[i][core]) {
+				ret = dma_multi_chan_domain_irq_register(
+						&dma_domain->data[i][j],
+						handler);
+				if (ret < 0)
+					return ret;
+
+				dma_domain->data[i][j].handler = handler;
+				dma_domain->data[i][j].arg = arg;
+
+				/* needed to unregister aggregated interrupts */
+				dma_domain->arg[i][core] =
+					&dma_domain->data[i][j];
+			}
+
+			interrupt_clear_mask(dma_domain->data[i][j].irq,
+					     BIT(j));
+
+			dma_interrupt(&dmas[i].chan[j], DMA_IRQ_UNMASK);
+
+			dma_domain->data[i][j].task = task;
+			dma_domain->channel_mask[i][core] |= BIT(j);
+
+			return 0;
+		}
+	}
+
+	return 0;
+}
+
+static void dma_multi_chan_domain_irq_unregister(struct dma_domain_data *data)
+{
+	interrupt_disable(data->irq, data);
+
+	interrupt_unregister(data->irq, data);
+}
+
+static void dma_multi_chan_domain_unregister(struct ll_schedule_domain *domain,
+					     uint32_t num_tasks)
+{
+	struct dma_domain *dma_domain = ll_sch_domain_get_pdata(domain);
+	struct dma *dmas = dma_domain->dma_array;
+	int core = cpu_get_id();
+	int i;
+	int j;
+
+	for (i = 0; i < dma_domain->num_dma; ++i) {
+		for (j = 0; j < dmas[i].plat_data.channels; ++j) {
+			/* channel not set as scheduling source */
+			if (!dma_is_scheduling_source(&dmas[i].chan[j]))
+				continue;
+
+			/* channel still running */
+			if (dmas[i].chan[j].status == COMP_STATE_ACTIVE)
+				continue;
+
+			/* channel owned by different core */
+			if (core != dmas[i].chan[j].core)
+				continue;
+
+			/* channel hasn't been running */
+			if (!(dma_domain->channel_mask[i][core] & BIT(j)))
+				continue;
+
+			dma_interrupt(&dmas[i].chan[j], DMA_IRQ_MASK);
+			dma_interrupt(&dmas[i].chan[j], DMA_IRQ_CLEAR);
+			interrupt_clear_mask(dma_domain->data[i][j].irq,
+					     BIT(j));
+
+			dma_domain->data[i][j].task = NULL;
+			dma_domain->channel_mask[i][core] &= ~BIT(j);
+
+			/* unregister interrupt */
+			if (!dma_domain->aggregated_irq)
+				dma_multi_chan_domain_irq_unregister(
+						&dma_domain->data[i][j]);
+			else if (!dma_domain->channel_mask[i][core])
+				dma_multi_chan_domain_irq_unregister(
+						dma_domain->arg[i][core]);
+
+			return;
+		}
+	}
+}
+
+static bool dma_multi_chan_domain_is_pending(struct ll_schedule_domain *domain,
+					     struct task *task)
```

Comment:
Can we have a small one line comment here for each function, makes it easier to follow.

Reply:
Done.